### PR TITLE
Remove binding redirect of Newtonsoft.Json from testhost config file

### PR DESCRIPTION
**Pull request by Faizan2304**
Issues: 
1) https://github.com/Microsoft/vstest/issues/391
2) https://github.com/Microsoft/vstest/issues/613

Fix: Remove binding redirect of Newtonsoft.json from testhost config file

### CODE REVIEW

File: scripts/build.ps1 [hunk 1]
```diff
@@ -219,15 +219,26 @@ function Publish-Package
     Write-Log "Package: Publish testhost.x86\testhost.x86.csproj"
     Publish-PackageInternal $testHostx86Project $TPB_TargetFramework $testhostFullPackageDir
 
-    # Copy over the Full CLR built testhost package assemblies to the $fullCLRPackageDir
-    Copy-Item $testhostFullPackageDir\* $fullCLRPackageDir -Force
+    # Binding redirection for Newtonsoft.Json.dll has been removed from testhost config file to resolve bug: https://github.com/Microsoft/vstest/issues/391
+    # Microsoft.TestPlatform.CommunicationUtilities depends on Newtonsoft.Json of version 8.0.3 and Microsoft.TestPlatform.CrossPlatEngine depends on Microsoft.Extensions.Dependencymodel which
+    # depends on Newtonsoft.Json of version 9.0.1. So at the end testhost has Newtonsoft.Json.dll of version 9.0.1 in their publish package.
+    # Since testhost uses CommunicationUtilities (and dont use that part of code of CrossPlatEngine which uses Newtonsoft.Json of version 9.0.1) so it needs Newtonsoft.Json.dll of version 8.0.3
```
Comment (AbhitejJohn) [inline scripts/build.ps1]: Might want to include why we made this change.

Comment (AbhitejJohn) [inline scripts/build.ps1]: Please also state that we have a very thin line on cross plat engine's dependency on newtonsoft.json 9.0.1(It actually should be just calling into 8.0.3 compatible APIs since that's what it finds at runtime.)

File: test/Microsoft.TestPlatform.Client.UnitTests/Microsoft.TestPlatform.Client.UnitTests.csproj [hunk 2]
```diff
@@ -10,6 +10,8 @@
     <TargetFrameworks>netcoreapp1.0;net46</TargetFrameworks>
     <AssemblyName>Microsoft.TestPlatform.Client.UnitTests</AssemblyName>
     <PackageTargetFallback Condition=" '$(TargetFramework)' == 'netcoreapp1.0' ">$(PackageTargetFallback);dnxcore50;portable-net45+win8</PackageTargetFallback>
+    <AutoGenerateBindingRedirects>true</AutoGenerateBindingRedirects>
+    <GenerateBindingRedirectsOutputType>true</GenerateBindingRedirectsOutputType>
```
Comment (AbhitejJohn) [inline test/Microsoft.TestPlatform.Client.UnitTests/Microsoft.TestPlatform.Client.UnitTests.csproj]: We need to figure out what this actually means.

**Review by AbhitejJohn — APPROVED** [review]

**Comment by Faizan2304** [comment]
@dotnet-bot test this please

File: test/Microsoft.TestPlatform.Client.UnitTests/Microsoft.TestPlatform.Client.UnitTests.csproj (hunk 2 above)
Comment (codito) [inline test/Microsoft.TestPlatform.Client.UnitTests/Microsoft.TestPlatform.Client.UnitTests.csproj]: @Faizan2304 why is this required for TP.Client.UnitTests? Can we add a comment please.

File: scripts/build.ps1 [hunk 3]
```diff
@@ -219,15 +219,29 @@ function Publish-Package
...
+    # if not provided any, so here it will use testhost config file which has binding redirect <bindingRedirect oldVersion="0.0.0.0-9.0.0.0" newVersion="9.0.0.0" />. Since the user's projects is dependent on Newtonsoft.Json
+    # of version 6.0.5 their tests break because of the redirection.
+    # In the TestPlatform repo: Microsoft.TestPlatform.CommunicationUtilities depends on Newtonsoft.Json(8.0.3) and Microsoft.TestPlatform.CrossPlatEngine depends on Microsoft.Extensions.Dependencymodel which
+    # depends on Newtonsoft.Json(9.0.1). So testhost ends up having Newtonsoft.Json.dll of version 9.0.1 in their publish package.
+    # Since testhost uses CommunicationUtilities (and does not use that part of CrossPlatEngine which uses Newtonsoft.Json of version 9.0.1) so it needs Newtonsoft.Json.dll of version 8.0.3 along with it.
+    # Hence, copying Newtonsoft.Json.dll of version 8.0.3 to test $testhostFullPackageDir. This is however very much dependent on the CrossPlatEngine code running in testhost to not call into APIs specific to 9.0.1.
```
Comment (codito) [inline scripts/build.ps1]: We're copying the JSON.net since CommunicationUtilities requires it. Suggest adding this root cause to the actual bug and refer it here for history.

After refactoring of test host providers, CrossPlatEngine no longer takes a dependency on JSON.net.

File: scripts/build.ps1 [hunk 4]
```diff
@@ -219,15 +219,29 @@ function Publish-Package
...
+
+    Write-Log "Package: copy file $newtonsoft to $testhostFullPackageDir"
+    $newtonsoft = Join-Path $env:TP_PACKAGES_DIR "newtonsoft.json\8.0.3\lib\net45\Newtonsoft.Json.dll"
+    Copy-Item $newtonsoft $testhostFullPackageDir -Force
+
+    # Copy over the Full CLR built testhost package assemblies to the Core CLR and Full CRL package folder.
```
Comment (codito) [inline scripts/build.ps1]: s/CRL/CLR

File: src/testhost.x86/testhost.x86.csproj [hunk 5]
```diff
@@ -14,6 +14,7 @@
   </PropertyGroup>
   <PropertyGroup Condition="'$(TargetFramework)' != 'netcoreapp1.0'">
     <RuntimeIdentifier>win7-x64</RuntimeIdentifier>
+    <AutoGenerateBindingRedirects>false</AutoGenerateBindingRedirects>
```
Comment (codito) [inline src/testhost.x86/testhost.x86.csproj]: Please add a comment why is this required. CrossPlatEngine no longer has a dependency on JSON.net.

File: src/testhost/testhost.csproj [hunk 6]
```diff
@@ -12,6 +12,7 @@
   </PropertyGroup>
   <PropertyGroup Condition="'$(TargetFramework)' != 'netcoreapp1.0'">
     <RuntimeIdentifier>win7-x64</RuntimeIdentifier>
+    <AutoGenerateBindingRedirects>false</AutoGenerateBindingRedirects>
```
Comment (codito) [inline src/testhost/testhost.csproj]: May be this is not required since CrossPlatEngine is not dependent on JSON.net.

File: test/Microsoft.TestPlatform.CommunicationUtilities.UnitTests/Microsoft.TestPlatform.CommunicationUtilities.UnitTests.csproj [hunk 7]
```diff
@@ -11,6 +11,8 @@
     <AssemblyName>Microsoft.TestPlatform.CommunicationUtilities.UnitTests</AssemblyName>
     <EnableCodeAnalysis>true</EnableCodeAnalysis>
     <PackageTargetFallback Condition=" '$(TargetFramework)' == 'netcoreapp1.0' ">$(PackageTargetFallback);dnxcore50;portable-net45+win8</PackageTargetFallback>
+    <AutoGenerateBindingRedirects>true</AutoGenerateBindingRedirects>
```
Comment (codito) [inline test/Microsoft.TestPlatform.CommunicationUtilities.UnitTests/Microsoft.TestPlatform.CommunicationUtilities.UnitTests.csproj]: Require comment.

Comment (codito) [inline test/Microsoft.TestPlatform.CommunicationUtilities.UnitTests/Microsoft.TestPlatform.CommunicationUtilities.UnitTests.csproj]: Same for all changes below.

**Review by codito — APPROVED** [review]
LGTM with few suggestions.

File: scripts/build.ps1 (hunk 3 above)
Comment (Faizan2304) [inline scripts/build.ps1]: Fixed in next commit

File: scripts/build.ps1 (hunk 4 above)
Comment (Faizan2304) [inline scripts/build.ps1]: Fixed in next commit

File: src/testhost/testhost.csproj (hunk 6 above)
Comment (Faizan2304) [inline src/testhost/testhost.csproj]: We will remove after testing once Mayank fix will go in

File: src/testhost.x86/testhost.x86.csproj (hunk 5 above)
Comment (Faizan2304) [inline src/testhost.x86/testhost.x86.csproj]: after PR https://github.com/Microsoft/vstest/pull/659, we may not need this.

**Comment by Faizan2304** [comment]
@dotnet-bot test this please

**Comment by Faizan2304** [comment]
@dotnet-bot test this please
